src/master/udp_rpc.hpp

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * udp_rpc.hpp - remote procedure call (RPC) over UDP transport module
+ */
+
+#ifndef _UDP_RPC_HPP_
+#define _UDP_RPC_HPP_
+
+#include "hostname.hpp"
+
+#include <functional>
+#include <map>
+#include <thread>
+
+namespace cubhb
+{
+
+  // type aliases
+  using socket_type = SOCKET;
+  using ipv4_type = std::uint32_t;
+  using port_type = std::uint16_t;
+
+  // constants
+  static const std::size_t BUFFER_SIZE = 4096;
+
+  enum message_type
+  {
+    UNKNOWN_MSG = 0,
+    HEARTBEAT = 1
+  };
+
+  // pack body and message type into the buffer
+  template <typename Body>
+  void pack (cubmem::extensible_block &buffer, message_type &type, const Body &body);
+
+  // unpack body and message type from the buffer
+  template <typename Body>
+  void unpack (const cubmem::block &buffer, message_type &type, Body &body);
+
+  class client_request
+  {
+    public:
+      client_request () = delete;
+      client_request (socket_type sfd, const cubbase::hostname_type &host, port_type port);
+
+      // send the request to destination host
+      void end () const;
+
+      // template function: set the payload of the client request
+      // Body will be automatically packed into a buffer
+      template <typename Body>
+      void set_body (message_type type, const Body &body);
```

Comment:
Please try  to clarify difference between message type, body and the packed message. Body sometimes refers to message content without type, and it sometimes refers to both type & content.

Reply:
Here `Body` is a template type, it could be any type used in overloaded function from `cubpacking::packer::pack_overloaded` or `cubpacking::unpacker::unpack_overloaded`.
In other cases, like: `client_request::m_body, server_response::m_body` and `server_request::m_body`, body is a packed message/buffer which will contain both `Body` and message type)

Reply:
fixed